### PR TITLE
Issue #206: PaymentController can only respond with 404 or with NullPointerException when creating Payment

### LINKED ISSUE #206: PaymentController can only respond with 404 or with NullPointerException when creating Payment
## 🐛 Describe the bug

The PaymentController currently calls to the method `SchedulePaymentApiMapper#enrichedSchedulePaymentRequestToPaymentDto(...)` immediately before it calls the PaymentService to persist the payment:

```java
return houseMember.flatMap(member -> adminOptional.filter(
        admin -> isUserAdminOfCommunityHouse(member.getCommunityHouse(), admin)))
        .map(admin -> schedulePaymentApiMapper.enrichedSchedulePaymentRequestToPaymentDto(
            schedulePaymentApiMapper.enrichSchedulePaymentRequest(request, admin,
                houseMember.get())))
        .map(paymentService::schedulePayment)
```

From here, the PaymentController attempts to persist the payment into the database. Before persisting the payment, it first tries to persist an Admin User.

From PaymentSDJpaService.java:
```java
  @Override
  public PaymentDto schedulePayment(PaymentDto request) {
    generatePaymentId(request);
    return createPaymentInRepository(request);
  }

  private PaymentDto createPaymentInRepository(PaymentDto request) {
    Payment payment = paymentMapper.paymentDtoToPayment(request);

    adminRepository.save(payment.getAdmin());
    paymentRepository.save(payment);

    return paymentMapper.paymentToPaymentDto(payment);
  }
```

But the problem is that the method `enrichedSchedulePaymentRequestToPaymentDto` is annotated this way:

```java
@Mappings({
      @Mapping(target = "admin", ignore = true),
      @Mapping(target = "member", ignore = true)
  })
  PaymentDto enrichedSchedulePaymentRequestToPaymentDto(
      EnrichedSchedulePaymentRequest enrichedSchedulePaymentRequest);
```

Because admin is marked as ignore, the property in the resulting PaymentDto will be null. Then when `createPaymentInRepository` is called to convert to Payment, it takes a null Admin and brings it over as null. Then `adminRepository.save(payment.getAdmin());` is called and the request fails with a NullPointerException.

## ⚠️ Current behavior

All valid requests fail with NPE.

## ✅ Expected behavior

Valid requests should persist the payment into the database with correct attributes.

## 💣 Steps to reproduce

Remove the `@Disabled` annotation from the Integration Test that I put in #205 and then run the test.

### PR DESCRIPTION
**Pull request by edwardUL99**
I didn't realise I had to name the commit with the specified pattern, sorry!

## 🚀 Description

This PR closes #206.

It has the following changes:

- An AfterMapping method has been added to SchedulePaymentApiMapper so that after mapping a payment request to an enriched one, it fills in the null admin and member which were causing the NullPointerExceptions.
- A ManyToOne relationship also had to be added to admin and member in Payment.java, as without them, there were issues with fields being too long
- Payment#description was defined as unique but this was causing issues after fixing the other issue. It makes sense that a description should not be unique

## 📄 Motivation and Context

This big fix resolves the NullPointerExceptions blocking pull request #205. It closes #206 

## 🧪 How Has This Been Tested?

Since pull request #205 has not been merged yet, I copied all the required code over to my branch to re-run the failing test. I removed the Disabled annotation and ran the test and it passed.

Also used Postman to test the scenario of creating a payment. I tried the scenario of creating a payment with an admin ID that is not in the community of the member ID and got the 404 response because it couldn't find that admin in the member's community.

## 📦 Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [X] My code follows the code style of this project(Do your best to follow code styles. If none apply just skip this).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: service/src/main/java/com/myhome/domain/Payment.java [hunk 1]
```diff
@@ -41,12 +44,14 @@
   private BigDecimal charge;
   @Column(nullable = false)
   private String type;
-  @Column(unique = true, nullable = false)
```
Comment (mslowiak) [inline service/src/main/java/com/myhome/domain/Payment.java:44]: nice catch!

File: service/src/main/java/com/myhome/controllers/mapper/SchedulePaymentApiMapper.java [hunk 2]
```diff
@@ -139,4 +123,19 @@ default EnrichedSchedulePaymentRequest enrichSchedulePaymentRequest(
         member.getName(),
         member.getCommunityHouse() != null ? member.getCommunityHouse().getHouseId() : "");
   }
+
+  default UserDto getEnrichedRequestAdmin(EnrichedSchedulePaymentRequest enrichedSchedulePaymentRequest) {
+    return UserDto.builder().userId(enrichedSchedulePaymentRequest.getAdminId())
```
Comment (mslowiak) [inline service/src/main/java/com/myhome/controllers/mapper/SchedulePaymentApiMapper.java]: Could you move `userId` to the next line?

File: service/src/main/java/com/myhome/controllers/mapper/SchedulePaymentApiMapper.java [hunk 3]
```diff
@@ -139,4 +123,19 @@ default EnrichedSchedulePaymentRequest enrichSchedulePaymentRequest(
...
+        .id(enrichedSchedulePaymentRequest.getAdminEntityId())
+        .name(enrichedSchedulePaymentRequest.getAdminName())
+        .email(enrichedSchedulePaymentRequest.getAdminEmail())
+        .encryptedPassword(enrichedSchedulePaymentRequest.getAdminEncryptedPassword())
+        .build();
+  }
+
+  default HouseMemberDto getEnrichedRequestMember(EnrichedSchedulePaymentRequest enrichedSchedulePaymentRequest) {
+    return new HouseMemberDto().id(enrichedSchedulePaymentRequest.getMemberEntityId())
```
Comment (mslowiak) [inline service/src/main/java/com/myhome/controllers/mapper/SchedulePaymentApiMapper.java]: Could you move `id` to the next line?

**Comment by mslowiak** [comment]
> A ManyToOne relationship also had to be added to admin and member in Payment.java, as without them, there were issues with fields being too long

@edwardUL99 
Can you elaborate on that? What means 'fields being too long'? 
BTW. Sorry for the late response on that

**Comment by edwardUL99** [comment]
> > A ManyToOne relationship also had to be added to admin and member in Payment.java, as without them, there were issues with fields being too long
> 
> @edwardUL99 
> Can you elaborate on that? What means 'fields being too long'? 
> BTW. Sorry for the late response on that

I don't really remember now, in hindsight should have documted the stack trace. I believe it was an exception with IDs being greater than 255 characters. Some reason adding ManyToOne fixed that, I'm not too sure.

**Review by mslowiak — CHANGES_REQUESTED** [review]
There are small changes but if you correct them we can merge that @edwardUL99

**Comment by mslowiak** [comment]
Thanks @edwardUL99

**Comment by edwardUL99** [comment]
@mslowiak The changes are done but got a codecov error. I tried rebasing but it said my branch diverged by 33 and 2 commits respectively between local and remote branch. I believe it is because the branch has already been pushed

**Comment by edwardUL99** [comment]
No problem @mslowiak

**Comment by mslowiak** [comment]
Feel free to take something new, there are plenty of new issues @edwardUL99 ! :)

**Pull request merged**